### ISSUE TITLE
Read Geb config only once and make it centrally accessible

### ISSUE THREAD

**Issue by kriegaex** [geb/issues#660]
The Geb configuration file should be read only once and then cached. Ideally, that would not happen on demand, e.g. when creating the first browser instance, but right when setting up the global Spock extension. Aside from increased efficiency and fewer file I/O, another advantage would be that configuration data could be used in `@Requires` and `@IgnoreIf` conditions without causing bootstrapping problems as lately described on the mailing list or in the Spock Gitter channel.

See the discussion started on the mailing list [here](https://groups.google.com/g/geb-user/c/mLyov0ZULL0/m/r_bDvLQCAQAJ). Quote:

> **Marcin Erdmann wrote:** Geb config is read whenever
> ConfigurationLoader.getConf() is called. It happens to be called via
> the parameterless constructor of Browser, which in turn is called for
> the first time getBrowser() is called on an instance of
> GebTestManager.
> 
> All that aside, yes, Geb config is definitely read more often than it
> needs to be. There will definitely be a performance penalty to parsing
> and executing that script for every single test but this has worked
> like that since forever (I started contributing to Geb in 2010 and
> it's how it was back then) so changing this will be a breaking change.
> It will be fairly easy to change to only load the config once per
> GebTestManager instance but that would be a breaking change so would
> need to be done only in Geb 6 which might be coming soon due to
> switching to Selenium 4 which is not backwards compatible.
> 
> If you feel like this would be worthwhile to change, Alexander, then
> please open an issue in the tracker and I will schedule it for Geb 6.
> 
>> **Alexander Kriegisch wrote:** I am not sure you are following the
>> Spock chat on Gitter. While talking to Jeremy there, I was looking
>> into how Geb actually works. Result (I wrote much more there): It is
>> a simple global Spock extension. But to my surprise, the Geb config
>> is not read just once during extension start, but quite often, e.g.
>> whenever the Geb test manager calls getBrowser in beforeTest, i.e.
>> for each feature method in a Geb spec. By printing callstacks from
>> the Geb config, this is easy to see.
>> 
>> So my question is: What is the reason for that? Are you expecting
>> cases in which the Geb config changes during a JVM session? The
>> driver configurations for multiple environments are closures, so they
>> can be evaluated without re-reading the file. Any system properties
>> defined there should also not depend on dynamic things like time
>> stamps, random values or whatever.
>> 
>>> **Marcin Erdmann wrote:** GebConfig.groovy is usually evaluated when
>>> geb.ConfigurationLoader.getConf() is executed which for example
>>> happens inside of the parameterless constructor for geb.Browser. If
>>> you want to have GebConfig.groovy evaluated early on as part of
>>> executing tests then one thing that comes to mind with the setup you
>>> describe is to put new ConfigurationLoader().conf in a static
>>> initalization block of BaseTest.

**Comment by kriegaex** [comment]
The absolute simplest way to do it would be something like this:

```patch
--- a/module/geb-core/src/main/groovy/geb/ConfigurationLoader.groovy	(revision Staged)
+++ b/module/geb-core/src/main/groovy/geb/ConfigurationLoader.groovy	(date 1649813968155)
@@ -34,6 +34,8 @@
  */
 class ConfigurationLoader {
 
+    final static Configuration CONFIG = new ConfigurationLoader().conf
+
     final String environment
     final Properties properties
     final GroovyClassLoader specialClassLoader
@@ -85,7 +87,7 @@
      * @see #getConfFromClass(String)
      */
     Configuration getConf() throws UnableToLoadException {
-        doGetConf(null) ?: doGetConfFromClass(null) ?: getDefaultConf()
+        CONFIG ?: doGetConf(null) ?: doGetConfFromClass(null) ?: getDefaultConf()
     }
 
     /**
```

That would instantly apply to alll these callers (the green ones are tests):

![image](https://user-images.githubusercontent.com/1537384/163083173-3c7831df-6afd-40dd-b9de-5c0b97405167.png)

Without tests:

![image](https://user-images.githubusercontent.com/1537384/163083350-7e2ba75d-a7be-4fd7-8fa5-f057ddb0b8cb.png)

Of course this can be made fancier or more configurable, e.g. by
  * allowing global caching to be deactivated via system property,
  * also using the cached instance when the other `getConf` methods with parameters are called,
  * not making `CONFIG` final and allow it to be re-initialised, e.g. by simply setting it to `null` and then re-assigning it the next time `getConf()` is called.

WDYT, @erdi?

**Comment by erdi** [comment]
FWIW, I don't think the caching mechanism should go into `ConfigurationLoader`, in my opinion it should go into `GebTestManager`. There might be use cases where you do want to re-parse the config and with what you suggest you are preventing such ability. Also, adding it to `GebTestManager` would affect smaller number of codepaths while solving the issue at hand of config being re-parsed unnecessarily for every single test.

I would hold off with the PR, this will be a simple enough change for me to apply and orchestrating this not being merged into master until Geb 6 would probably be more trouble than it's needed. Unless you really feel like contributing this and you agree with my suggestion of adding this to `GebTestManager` then I can create a `geb-6` branch which you could target.

Secondly, I don't think that you should depend on ordering and parse the config eagerly like you suggest. I think using a config script to change global state when it's parsed and then depend on that global state elsewhere is a hack and this could be solved in a better way. I don't want to encourage this pattern and hence would like to only parse the script when the first instance of `Browser` is requested.

**Comment by kriegaex** [comment]
> I don't think the caching mechanism should go into `ConfigurationLoader`, in my opinion it should go into `GebTestManager`.

That depends on the goal. Mine was to make the default behaviour (caching activated) effective for as many code paths as possible, whereas your goal seems to be to limit it. Different goal, different target class. If my goal was yours, I would agree. But that is not the case.

> There might be use cases where you do want to re-parse the config

Please name a few good examples.

> with what you suggest you are preventing such ability

No, I am not. I explicitly mentioned that it should be possible to globally deactivate caching via system property. Furthermore, my comment was a proof-of-concept style sketch, so we know where to start. It would be easy to overload by `getConf(boolean useCache)`, if anyone wants to enforce loading on a case-by-case basis. Even without that, the other overloaded methods (which are not caching anything) already exist and can be used as workarounds. But probably, the overload with the boolean is the simplest and most transparent way for every user to get what they want, if they need anything other than the default. This is a standard way of designing an API.

> Also, adding it to `GebTestManager` would affect smaller number of codepaths while solving the issue at hand of config being re-parsed unnecessarily for every single test.

Like I said, I want to make caching by default as broad as possible, so Geb does **not** have different defaults in different situations.

> I would hold off with the PR, this will be a simple enough change for me to apply and orchestrating this not being merged into master until Geb 6 would probably be more trouble than it's needed.

Then why not integrate it now and reverse the default for the system property in release 6? That way, users could opt in and provide feedback starting with the next minor 5.x release. We could add tests etc., and you would not have to hold off anything - the classical feature toggle strategy.

> I don't think that you should depend on ordering and parse the config eagerly like you suggest. I think using a config script to change global state when it's parsed and then depend on that global state elsewhere is a hack and this could be solved in a better way.

Actually, I do not understand your comment. There has been Geb config for ages, this is not changing. It will just be read a little earlier than before, in order to be available to other extensions at a time when currently it is not. Isn't the Geb configuration meant to be global for the extension's life time? If not, the Book of Geb does not explain that. Every user probably assumes  that it is global during the session, and I cannot imagine that anyone assumes it will be read multiple times.

> I don't want to encourage this pattern 

Which pattern? I really don't get your point.

> hence would like to only parse the script when the first instance of `Browser` is requested.

That is **too late,** because the browser is requested **after** `@IgnoreIf` or `@Requires` conditions have been evaluated. That is the whole point of raising this issue. Remember what I wrote on the mailing list.

**Comment by erdi** [comment]
> Mine was to make the default behaviour (caching activated) effective for as many code paths as possible, whereas your goal seems to be to limit it

I'm coming at it from a perspective of a maintainer of a framework that's been around for while and for which stability is important thus the less breaking changes the better. Personally I'd rather we limit the new behaviour to where the problem we are trying to solve is (which imho is loading the config before every test) and thus limit the amount of change.

> Please name a few good examples.

Being able to load config for different environments in the same JVM comes to mind, for example.

> I explicitly mentioned that it should be possible to globally deactivate caching via system property.

I think that would make it more complex that it needs to be. I'm just not sure that caching it by default is the right default.

> so Geb does not have different defaults in different situations

I don't see it as different defaults in different situations but that's probably a matter of perspective. 

> Then why not integrate it now and reverse the default for the system property in release 6?

Yeah, we could feature flag it, that's a good point.

> That way, users could opt in and provide feedback starting with the next minor 5.x release.

I doubt anybody apart form you would, but that's a different matter.

> It will just be read a little earlier than before, in order to be available to other extensions at a time when currently it is not.

That's my point. It's bad design to depend on when the config script is executed to setup some global state (system properties) and then use that global state from another extension point. That's the pattern I don't want to encourage. It's brittle and the brittleness of it is the reason why we're actually discussing it. Currently the script is loaded the first time a `Browser` instance is instantiated by `GebTestManager` because this is the point at which it's needed and I'd prefer for it to stay that way. The config script should imho be solely used for configuring the `Browser` instance and not to set some environment vars.

Anyway, while I don't agree with your views I don't feel strongly enough about this to oppose. Please feel free to open a PR with your suggested implementation and I will accept it in principle.

**Comment by kriegaex** [comment]
@erdi: Thank you so much for taking care of this issue.

Quick question, because I did not read the whole diff or test your implementation: Is GebConfig now read soon enough to be used in `@IgnoreIf` or `@Requires` conditions?

**Comment by erdi** [comment]
No, it's not as I still don't see a reason why it should and don't believe that depending on when it's read in your `@IgnoreIf` or `@Requires` conditions to be a good engineering practice.

**Comment by erdi** [comment]
You can quite easily force it to with a base class if you really need to
```
class EagerConfigGebSpec extends Specification {

    private final static Configuration CONFIG = new ConfigurationLoader().conf

    private final static GebTestManager TEST_MANAGER = new SpockGebTestManagerBuilder()
        .withBrowserCreator { new Browser(CONFIG) }
        .build()

    GebTestManager getTestManager() {
        TEST_MANAGER
    }

}
```
Another option would be to load it in a global Spock extension.

**Comment by kriegaex** [comment]
The whole point of this issue was to be able to use properties defined in Geb config in Spock execution conditions intuitively. That was the origin of the discussion on Gitter, if you remember. I totally disagree with the "good engineering practice" argument. Users can expect that a **global** extension like Geb is fully initialised before using an annotation-driven one like `@IgnoreIf` or `@Requires`. IMO, this is just not good engineering practice to violate that expectation.

The requirement to implement yet another global extension in order to fix this one seems weird to me. So does the idea to define a base class and change the whole inheritance hierarchy for something that simple, just so as to make Geb work the way every normal (as in non-expert) user would expect it to. Don't make your users jump higher than necessary. Why not walk the last mile? The config is loaded just once now. You could easily do that when initialising the global Geb extension.

**Comment by erdi** [comment]
> The whole point of this issue was to be able to use properties defined in Geb config in Spock execution conditions intuitively. That was the origin of the discussion on Gitter, if you remember.

Yes I do remember that

> I totally disagree with the "good engineering practice" argument.

Looks like we have already established that we don't agree about this.

> Users can expect that a global extension like Geb is fully initialised before using an annotation-driven one like `@IgnoreIf` or `@Requires`.

I'm not sure what "full initialisation" of Geb you are referring to, I don't recognise that as a concept. Also, Geb and `@IgnoreIf` or `@Requires` are completely unrelated.

> IMO, this is just not good engineering practice to violate that expectation.

**Comment by erdi** [comment]
> The whole point of this issue was to be able to use properties defined in Geb config in Spock execution conditions intuitively. That was the origin of the discussion on Gitter, if you remember. 

Yes, I do remember.

> I totally disagree with the "good engineering practice" argument

Looks like we have established that we disagree on this.

> Users can expect that a global extension like Geb is fully initialised before using an annotation-driven one like `@IgnoreIf` or `@Requires`.

Not sure what you refer to as "fully initialised" and why executing Geb's configuration script would be part of that, I don't recognise this as a concept in the project.

> IMO, this is just not good engineering practice to violate that expectation.

Not all expectations have to be correct. I don't believe this one is.

> The requirement to implement yet another global extension in order to fix this one seems weird to me.

There is nothing broken in Geb with regards to that, in my opinion.

>  just so as to make Geb work the way every normal (as in non-expert) user would expect it to

I'm not sure what makes you write statements like this. You made up your mind that there is a "right" point in Spock's lifecycle for Geb's configuration script to be executed. You are trying to impose that there should be a contract about this. I can tell you that there never was and never will be one simply because these two things are completely separate and independent as far as I'm concerned.

I think this had enough discussion already and therefore I'm not going to discuss it anymore. Thanks.
